src/main/java/com/minecolonies/coremod/client/gui/WindowResearchTree.java

```diff
@@ -137,299 +199,775 @@ public void onButtonClicked(@NotNull final Button button)
       final ZoomDragView view,
       final List<String> researchList,
-      final ILocalResearchTree tree,
-      final boolean parentResearched,
-      final boolean abandoned,
-      final int parentHeight)
+      final boolean abandoned)
     {
         int nextHeight = height;
         for (int i = 0; i < researchList.size(); i++)
         {
             final String researchLabel = researchList.get(i);
-            int offsetX = (depth * (GRADIENT_WIDTH + X_SPACING));
 
             final IGlobalResearch research = IGlobalResearchTree.getInstance().getResearch(branch, researchLabel);
-            final ILocalResearch localResearch = tree.getResearch(branch, research.getId());
-            final ResearchState state = localResearch == null ? ResearchState.NOT_STARTED : localResearch.getState();
-            final IGlobalResearch parentResearch = IGlobalResearchTree.getInstance().getResearch(branch, research.getParent());
+            if (research.isHidden() && !IGlobalResearchTree.getInstance().isResearchRequirementsFulfilled(research.getResearchRequirement(), this.building.getColony()))
+            {
+                continue;
+            }
+
+            final int offsetX = (depth * (GRADIENT_WIDTH + X_SPACING));
+            final int offsetY = (nextHeight + Math.min(i, 1)) * (GRADIENT_HEIGHT + Y_SPACING) + Y_SPACING + TIMELABEL_Y_POSITION;
 
-            boolean trueAbandoned = abandoned;
-            if (depth != 0 && abandoned == false && state == ResearchState.NOT_STARTED && parentResearch.hasResearchedChild(tree) && parentResearch.hasOnlyChild())
+            final boolean trueAbandoned = drawResearchItem(view, offsetX, offsetY, research, abandoned);
+
+            if (!research.getParent().isEmpty())
             {
-                trueAbandoned = true;
+                drawArrows(view, offsetX + INITIAL_X_OFFSET - X_SPACING, offsetY, researchList, research.getParent(), i, nextHeight, height);
             }
 
-            final Gradient gradient = new Gradient();
-            gradient.setSize(GRADIENT_WIDTH, GRADIENT_HEIGHT);
-            gradient.setPosition(offsetX + INITIAL_X_OFFSET, (nextHeight + Math.min(i, 1)) * (GRADIENT_HEIGHT + Y_SPACING) + Y_SPACING);
-            if (state == ResearchState.IN_PROGRESS)
+            if (!research.getChildren().isEmpty())
             {
-                gradient.setGradientStart(227, 249, 184, 255);
-                gradient.setGradientEnd(227, 249, 184, 255);
-                view.addChild(gradient);
+                nextHeight =
+                  drawTree(nextHeight + Math.min(i, 1), depth + 1, view, research.getChildren(), trueAbandoned);
             }
-            else if (trueAbandoned && state != ResearchState.FINISHED)
+            else
             {
-                gradient.setGradientStart(191, 184, 172, 255);
-                gradient.setGradientEnd(191, 184, 172, 255);
-                view.addChild(gradient);
+                nextHeight += Math.min(i, 1);
```

Comment:
shouldn't this be +=1?
I'm not sure if nextHeight is count of rows or Y offset in pixels, but this code seems a bit of both to me